### PR TITLE
Refactor ZeRO configs to use Pydantic

### PR DESCRIPTION
**Pull request by mrwyattii**
Brings Pydantic into the zero config:
- Defines a base `DeepSpeedConfigModel` class which ZeRO and offload configs inherit. This class allows us to enforce extra params are not provided in the config and also implements logic for deprecated config params
- Unify the ZeRO (and offload) `constants.py` and `config.py` into a single Pydantic class that contains the default values for all parameters
- Changes in the rest of the code to remove values defined in `constants.py`

### CODE REVIEW

**Comment by jeffra** [comment]
This PR is aiming to help with this longstanding issue https://github.com/microsoft/DeepSpeed/issues/653

/cc @stas00

**Comment by stas00** [comment]
Wonderful! Thank you for the heads up, Jeff!

Also probably extending hjson-support to the rest of DS config would be great, but doesn't have to be part of this PR - just a suggestion.

**Comment by jeffra** [comment]
Absolutely, I think adding hjson + pydantic support is the eventual goal. One all of our configs are managed by pydantic the goal is to also replace the [config-json](https://www.deepspeed.ai/docs/config-json/) page with a [read-the-docs style page](https://deepspeed.readthedocs.io/en/latest/) that is auto-generated from doc-strings that are easier to maintain and more accurate. However, one step at a time :)

**Comment by rocm-mici** [comment]
Can one of the admins verify this patch?

File: deepspeed/runtime/config.py [hunk 1]
```diff
@@ -864,7 +851,13 @@ def _initialize_params(self, param_dict):
         self.gradient_predivide_factor = get_gradient_predivide_factor(param_dict)
         self.sparse_gradients_enabled = get_sparse_gradients_enabled(param_dict)
 
-        self.zero_config = DeepSpeedZeroConfig(param_dict)
+        if ZERO_OPTIMIZATION in param_dict:
```
Comment (tjruwase) [inline deepspeed/runtime/config.py:854]: Refactoring this logic could make it easier to read. Something like 
`self.zero_config = get_...`

File: deepspeed/autotuning/autotuner.py [hunk 2]
```diff
@@ -499,7 +498,7 @@ def tune_space(self,
                    prev_best_mbs=0,
                    prev_best_metric_val=0):
         config_zero = tuning_space.get(ZERO_OPTIMIZATION, {})
-        stage = config_zero.get(ZERO_OPTIMIZATION_STAGE, ZERO_OPTIMIZATION_STAGE_DEFAULT)
+        stage = config_zero.stage
```
Comment (tjruwase) [inline deepspeed/autotuning/autotuner.py:501]: Isn't this a bug if `config_zero` equals `{}`, and leads to evaluation of `{}.stage`?

File: deepspeed/runtime/config_utils.py [hunk 3]
```diff
@@ -8,6 +8,86 @@
...
+from deepspeed.utils import logger
+
+
+class DeepSpeedConfigModel(BaseModel):
+    """
+    This class should be used as a base for all DeepSpeed configs. It extends
+    pydantic.BaseModel to allow for deprecated fields. To enable this feature,
+    add deprecated=True to pydantic.Field:
+
+    my_dep_field: int = Field(0, deprecated=True)
```
Comment (tjruwase) [inline deepspeed/runtime/config_utils.py:21]: Can you document first argument of `Field(`, i.e., `0`.?

**Review by tjruwase — APPROVED** [review]

**Pull request merged**